File: 1011_capacityToShipPackagesWithinDDay/main.py
from typing import List

class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        l, r = max(weights), sum(weights)

        while l < r:
            m = (l + r) // 2
            cur = 0
            daysNeeded = 1
            for w in weights:
                if cur + w > m:
                    daysNeeded += 1
                    cur = 0
                cur += w
            if daysNeeded > days:
                l = m + 1
            else:
                r = m
        return l

File: 1011_capacityToShipPackagesWithinDDay/main.cpp
#include <vector>
#include <algorithm> // max_element
#include <numeric> // accumulate
class Solution {
public:
    int shipWithinDays(std::vector<int>& weights, int days) {
        int l = *std::max_element(weights.begin(), weights.end());
        int r = std::accumulate(weights.begin(), weights.end(), 0);
        int m, dayNeeded, curr;
        while(l < r) {
            m = (l + r) / 2;
            dayNeeded = 1;
            curr = 0;
            for(int w : weights) {
                if(curr + w > m) {
                    ++dayNeeded;
                    curr = 0;
                }
                curr = curr + w;
            }
            if(dayNeeded > days)
                l = m + 1;
            else
                r = m;
        }
        return l;
    }
};